modules/aggregation-layers/src/aggregation-layer.js

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import {CompositeLayer, AttributeManager, experimental} from '@deck.gl/core';
+import {cssToDeviceRatio, log} from '@luma.gl/core';
+const {compareProps} = experimental;
+
+// props when changed results in new uniforms that requires re-aggregation
+const UNIFORM_PROPS = [
```

Comment:
Layers should not be aware of the implementation detail of extensions. This is not going to scale when we support other extensions.

Reply:
Yeah, thought about it, ideal solution for this would be the extension itself expose these props and our existing prop diffing, differentiate these prop changes with existing `changeFlags.propsChanged`, so that aggregation is only run when these props are changed (not when any prop changed).
We don't need this for non composite layers, because we render the layers any time a prop is changed, but that is un-necessary and expensive for aggregation layers. I can create a task for this.

Reply:
Maybe you can add a `LayerExtension.needsRedraw()` method for this purpose?

Reply:
Will do this in separate PR #3844